Optional compatibility

dukhovnyi opened the pull request:
Hi,
I've used the framework a little bit and found some inconsistence that could be fixed. Please @blootsvoets review this PR.
Adds to AvroValueConvertible compatibility with Optional type and corrects Array extension.

dukhovnyi committed:
Adds to AvroValueConvertible compatibility with Optional type and corrects Array extension.

blootsvoets commented on Sources/BlueSteel/AvroValueConvertible.swift:
```diff
@@ -76,13 +76,13 @@ extension Data: AvroValueConvertible {
 }
 
-extension Dictionary: AvroValueConvertible where Key == String, Value == AvroValueConvertible {
+extension Dictionary: AvroValueConvertible where Key == String, Value: AvroValueConvertible {
```
I recall also having used this syntax and in an earlier Swift version at least it did not work for me. I suppose the tests pass though?

dukhovnyi replied:
Could be, but in the latest Swift version Xcode throws the error `Referencing instance method 'toAvro()' on 'Array' requires the types 'String' and 'AvroValueConvertible' be equivalent`.
In my opinion, usage conforming `Value` to `AvroValueConvertible` protocol, in this case, is more convenient. What do you think?

dukhovnyi replied:
Yes, tests pass though but probably I should add a few additional tests to fix this behavior and be sure that this extension works properly for Arrays and Dictionaries. I'll do it soon :)

blootsvoets commented on Sources/BlueSteel/coding/JsonAvroDecoder.swift:
```diff
@@ -61,10 +61,10 @@ open class JsonAvroDecoder : AvroDecoder {
                     .filter { result.keys.contains($0.name) }
-                    .map { field in (field.name, try decode(any: result[field.name]!, as: field.schema)) }
-                return Dictionary(uniqueKeysWithValues: fields)
+                    .reduce(into: [:]) { $0[$1.name] = try decode(any: result[$1.name]!, as: $1.schema) }
```
Could you please use named variables here for clarity? In that case the `fields` variable does not need to be used, the result can just directly be returned.

dukhovnyi replied:
Sure, I'll update this code and use named variables here.

>  In that case the fields variable does not need to be used, the result can just directly be returned.

I'd like to propose you don't remove `fields` because in debug time we can put a breakpoint on line with `return field` and check/edit value if needed. So it is not so important but pretty useful :) What do you think about it?

blootsvoets reviewed (commented):
Thanks a lot for your contribution. I have two questions, otherwise I’m happy to merge and release this.

dukhovnyi commented:
@blootsvoets I updated the PR and added test, so please review this PR.

I added test to be sure that Array with `String` items and array with `AvroValueConvertible` items are valid for compilation and extension for Optional works correctly.

blootsvoets reviewed (approved):
Looks good to me. Thanks again for your contribution!
